Query endpoint

abarciauskas-bgse opened the pull request:
@jflasher opening this PR to determine next steps. Not sure if / when I will get to do more parquet work. I was able to generate parquet using glue, but it's not query-optimized yet.

Wondering if we should still incorporate the query endpoint as part of the production deployment or just a local option. What do you think?

jflasher commented:
Thanks @abarciauskas-bgse! Yeah, I think we can get this in place without having the Parquet/ORC conversion figured out. We'll need to update the docs and when we do, I think we should call this out as experimental or similar. Basically, if we find that opening this up to people starts taking too many resources, we'll likely want to severely limit it or shut it down as a public endpoint altogether. 

I'll try to take a look at this in the near future but just from a quick glance, I think we'll need to include something that makes sure we only ever have `X` number of these requests active at one time and the user should just be told to try again later I guess? Also, and this one is on me, I think we'll need to have these queries get sent to a new public bucket so that the URL will work when people receive it. Our current Athena bucket is not publicly available.

abarciauskas-bgse commented:
@jflasher Thanks for taking a look. Where can I update the docs? Happy to do that.

Regarding limiting the number of concurrent queries, AWS already does this so I'm wondering if it will be sufficient. I can't find a way to do it ourselves: The Athena API doesn't provide a way to get the number of `STATE = "RUNNING"` queries without fetching all queries and iterating. I.e. [listQueryExecutions](https://docs.aws.amazon.com/AWSJavaScriptSDK/latest/AWS/Athena.html#listQueryExecutions-property) doesn't allow you to filter on query state and I don't see another appropriate method to fetch the information we need.

However, there is a limit set by AWS, see [this documentation](https://docs.aws.amazon.com/athena/latest/ug/service-limits.html). In testing, I found the limit to be closer to 30 than 20 ¯\\_(ツ)_/¯.

When running the following script (against local server):
```javascript
const http = require('http');
const aws = require('aws-sdk');
const athena = new aws.Athena({region: 'us-east-1'});

var responses = [];
var queryExecutionIds = [];
const times = 30;
const openaq_url = 'http://localhost:3004/v1/query';

try {
  for (var i=0; i < times; i++) {
    http.get(openaq_url, (res) => {
      res.setEncoding('utf8');
      var rawData = '';
      res.on('err', (err) => console.log(`Got error: ${err}`))
      res.on('data', (chunk) => { rawData += chunk; });
      res.on('end', () => {
        try {
          const parsedData = JSON.parse(rawData);
          console.log(`responses status code: ${res.statusCode}`)
          console.log(parsedData);
          const queryId = parsedData.results.queryId;
          queryExecutionIds.push(queryId);
          if (i === times) {
            queryExecutionIds.forEach((id) => {
              athena.stopQueryExecution({ QueryExecutionId: id }, (err, res) => {
                console.log('Stopping query');
              });
            });         
          }          
        } catch (e) {
          console.error(e.message);
        }
      });
    });
  };
} catch (e) {
  console.log(e)
};
```

I consistently see 1 of these errors:

```json
{
  "message": "Rate exceeded",
  "code": "ThrottlingException",
  "time": "2018-12-29T21:51:45.931Z",
  "requestId": "a5d76183-d97d-434e-83ea-bd3d1c8650f1",
  "statusCode": 400,
  "retryable": true
}
```
so I added [this code](https://github.com/openaq/openaq-api/pull/387/commits/c3e931e355187c90fd1a7ceee171996a7e4bdd44)

I can do more of these tests if it helps. Let me know if you have any ideas about unit testing this feature - since `tests.js` uses a test server it seems like it would be impossible to stub `athena.startQueryExecution` so the best I think the best I can do in the current framework is to test using a script like the one in this comment.

I changed the env var for storing query output to `ATHENA_QUERY_OUTPUT_BUCKET`. I'm assuming this bucket is something you can help create and configure when we are ready for deployment.

jflasher commented:
@abarciauskas-bgse thanks for the info! Docs are inline and generated automatically from there, you can see https://github.com/openaq/openaq-api/blob/develop/api/routes/countries.js#L8 as an example. You should be able to run `npm run docs` locally if you want to see what the output looks like.

The reason I think we need to figure out the concurrency issue is that our system always needs to have 3 available to do the aggregations. So if the account limit is X, we need to limit this endpoint to only use X - 3 to leave those ones for the system 3. We always want those aggregations queries to be able to run. Do you think there is any problem with fetching all the queries and iterating over them? I'm presuming it's just returning metadata so hopefully isn't a big/long request? 

And yep, I can help with the final query bucket when it comes to deployment time!

abarciauskas-bgse commented:
Thanks @jflasher - just noting that I will follow up on this this weekend at the latest

abarciauskas-bgse commented:
@jflasher ok I've added a check for concurrent queries (`canRunQuery`) and some tests.

Let me know when you have time to review!

jflasher commented:
Thanks @abarciauskas-bgse, I'm going to try and look at this soon and see how it works actually connected to Athena + S3. I think I'll need to hook up some other pieces to make sure the files are actually accessible once they show up. That will live outside anything here I think.

jflasher commented:
Just a note I’m holding on this pending the stationID work as it many impact how we search for items in general.  


On January 21, 2019 at 01:14:57, Aimee Barciauskas (notifications@github.com(mailto:notifications@github.com)) wrote:

>  
> @jflasher(https://github.com/jflasher) ok I've added a check for concurrent queries (canRunQuery) and some tests.
>  
>  
> Let me know when you have time to review!
>  
>  
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub(https://github.com/openaq/openaq-api/pull/387#issuecomment-455959666), or mute the thread(https://github.com/notifications/unsubscribe-auth/AAz0JoyqYqMK43Q5upByrjLplJ6yB2Meks5vFVrhgaJpZM4ZfUXt).
>  
>  
>

vgeorge commented:
Work on unique location id should be followed up at https://github.com/openaq/openaq-api/pull/394 now.

jflasher commented:
Reopening this up as it's for different work, unless I am misunderstanding something.

abarciauskas-bgse commented:
Closed as this was stale

russbiggs commented:
@JouzefB Thanks for taking the time to develop this pull request. In general new features are best discussed beforehand in github issues to ensure they are in alignment and technically feasible. I would also prefer that the PR only include updates for a single bug fix or feature to keep the discussion and edits encapsulated. This is relevant because the 429 issue is already being addressed in a separate PR #384 based on #386.  

If you are still interested in contributing issue #300 or #378 may be good issues, but would likely need a little discussion to ensure the full extent of the problem is clear, they are both a bit obtuse. If you are interested in one of those comment on the issue and I can assign it to you.
